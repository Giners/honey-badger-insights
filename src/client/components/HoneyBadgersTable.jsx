import React from 'react'
import PropTypes from 'prop-types'

import {
  PagingState,
  IntegratedPaging,
  RowDetailState,
  SearchState,
  IntegratedFiltering,
} from '@devexpress/dx-react-grid'

import {
  Grid,
  Table,
  TableHeaderRow,
  TableRowDetail,
  PagingPanel,
  Toolbar,
  SearchPanel,
} from '@devexpress/dx-react-grid-material-ui'

import HoneyBadgerDetails from './HoneyBadgerDetails'

/**
 * Takes a list of honey badgers and their details and turns it into rows of details for use in the
 * table that displays the details about honey badgers. It presumes that at a minimum that the
 * IP address ('ipAddress') and the amount of times the honey badger has been seen in the last 24
 * hours ('count') are present on the details.
 */
const getHoneyBadgersRows = honeyBadgers =>
  honeyBadgers.map(honeyBadger => {
    // Supply the minimum amount of info we expect in a row which is the IP address and the amount
    // of times we have seen the entity in the last 24 hours. Add a 'Loading' flag as the default
    // for all of the other info that may not be loaded yet.
    const { ipAddress, count, geoLocation, as } = honeyBadger

    const row = {
      ipAddress,
      count,
      country: 'Loading',
      asn: 'Loading',
    }

    // Add the additional info if it has been loaded:
    // * Autonomous system number if autonomous system info has been loaded
    // * Country of origin if geospatial location info has been loaded
    if (geoLocation) {
      row.country = geoLocation.country
    }

    if (as) {
      row.asn = as.asn
    }

    return row
  })

const HoneyBadgersTable = ({ honeyBadgers }) => {
  const renderHoneyBadgerDetails = ({ row: { ipAddress } }) => (
    <HoneyBadgerDetails honeyBadger={honeyBadgers.get(ipAddress)} />
  )

  renderHoneyBadgerDetails.propTypes = {
    row: PropTypes.shape({
      ipAddress: PropTypes.string.isRequired,
    }).isRequired,
  }

  return (
    <Grid
      rows={getHoneyBadgersRows([...honeyBadgers.values()])}
      columns={[
        { name: 'ipAddress', title: 'IP Address' },
        { name: 'count', title: 'Times Seen Last 24 Hours' },
        { name: 'country', title: 'Country of Origin' },
        { name: 'asn', title: 'Autonomous System #' },
      ]}
    >
      <RowDetailState />
      <SearchState />
      <IntegratedFiltering />
      <PagingState defaultCurrentPage={0} pageSize={5} />
      <IntegratedPaging />
      <Table />
      <TableHeaderRow />
      <TableRowDetail contentComponent={renderHoneyBadgerDetails} />
      <Toolbar />
      <SearchPanel />
      <PagingPanel />
    </Grid>
  )
}

HoneyBadgersTable.propTypes = {
  honeyBadgers: PropTypes.instanceOf(Map).isRequired,
}

export default HoneyBadgersTable
