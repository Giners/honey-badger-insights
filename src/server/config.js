/**
 * Config object that contains information/details to allow the app to function properly.
 * Information/details are gathered from environment variables. If the appropriate environment
 * variable isn't set then a default value is used. Note that some default values will cause the app
 * to no function properly, such as defaults for credentials for services that are used.
 */
const config = {
  // Config for the app/server
  app: {
    port: process.env.port || 3000,
  },
  // Service credentials
  serviceCreds: {
    honeyDB: {
      id: process.env.HONEYDB_API_AUTH_ID || 'No API auth ID found!',
      key: process.env.HONEYDB_API_AUTH_KEY || 'No API auth key found!',
    },
  },
}

export default config
