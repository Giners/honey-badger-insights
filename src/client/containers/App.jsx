import 'isomorphic-fetch'
import { ApolloClient } from 'apollo-client'
import { HttpLink } from 'apollo-link-http'
import { InMemoryCache } from 'apollo-cache-inmemory'

import React from 'react'

import HoneyBadgersTable from './../components/HoneyBadgersTable'
import {
  autonomousSystemsQuery,
  geoLocationsQuery,
  topHoneyBadgersQuery,
} from './../graphql/queries'
import config from './../../server/config'

const apolloClient = new ApolloClient({
  link: new HttpLink({
    uri: `${config.app.hostURL}:${config.app.port}${
      config.serviceURIs.graphQL
    }`,
  }),
  cache: new InMemoryCache(),
})

/** The initial state of the App component when it is mounted. Exported for testing purposes. */
export const initialState = { honeyBadgers: new Map(), error: null }

/**
 * The <App> component is a stateful React component that is responsible for providing the "root"
 * for our visual and logical React components that make up our app. Its most important job is to
 * query for information about honey badgers and then to pass that information along to child
 * components so they can render it into a meaningful visualization.
 */
class App extends React.Component {
  /**
   * Helper method that makes a query to the GraphQL service/API endpoint for the top honey badgers
   * that have been seen in the past 24 hours.
   *
   * @return {Promise<Map<string, HoneyBadger>, error>} Returns a promise where it is resolved to a
   * map that contains info about the top honey badgers. The map is keyed off of the honey badgers
   * IP address. Should this fail to query the GraphQL service/API endpoint or for any other reason
   * this promise will be rejected with the error.
   */
  static async collectTopHoneyBadgersInfo() {
    const { topHoneyBadgers } = (await apolloClient.query({
      query: topHoneyBadgersQuery,
    })).data

    const thbKVPs = topHoneyBadgers.map(topHoneyBadger => [
      topHoneyBadger.ipAddress,
      topHoneyBadger,
    ])
    return new Map(thbKVPs)
  }

  /**
   * Helper method that given a map of honey badgers where they are keyed off of their IP address
   * will make a query to the GraphQL service/API endpoint for said honey badger for info about the
   * autonomous systems that is responsible/in charge of said honey badger.
   *
   * Note that this function isn't pure and has side effects in that it modifies the map of honey
   * badgers passed into it to add in the info about the autonomous systems that has been
   * collected.
   *
   * @param {Map<string, HoneyBadger>} honeyBadgers - Map of honey badgers where they are keyed off
   * of their IP address
   *
   * @return {Promise<undefined, error>} Returns a promise where it will resolve if able to
   * successfully query for info about autonomous systems as well as add the info to the map of
   * honey badgers that is passed into it. Should we fail to query the GraphQL service/API endpoint
   * or for any other reason this promise will be rejected with the error.
   */
  static async collectAutonomousSystemsInfo(honeyBadgers) {
    const ipAddresses = [...honeyBadgers.keys()]

    const { autonomousSystems } = (await apolloClient.query({
      query: autonomousSystemsQuery,
      variables: { ipAddresses },
    })).data

    // Join the data about the autonomous system that the honey badger falls under with the
    // existing data about honey badgers
    autonomousSystems.forEach(({ ipAddress, ...rest }) => {
      const honeyBadger = honeyBadgers.get(ipAddress)

      honeyBadgers.set(ipAddress, { ...honeyBadger, as: { ...rest } })
    })
  }

  /**
   * Helper method that given a map of honey badgers where they are keyed off of their IP address
   * will make a query to the GraphQL service/API endpoint for said honey badger for info about
   * where said honey badger is located geospatially.
   *
   * Note that this function isn't pure and has side effects in that it modifies the map of honey
   * badgers passed into it to add in the info about the locations that has been collected.
   *
   * @param {Map<string, HoneyBadger>} honeyBadgers - Map of honey badgers where they are keyed off
   * of their IP address
   *
   * @return {Promise<undefined, error>} Returns a promise where it will resolve if able to
   * successfully query for info about the locations as well as add the info to the map of honey
   * badgers that is passed into it. Should we fail to query the GraphQL service/API endpoint
   * or for any other reason this promise will be rejected with the error.
   */
  static async collectGeoLocationsInfo(honeyBadgers) {
    const ipAddresses = [...honeyBadgers.keys()]

    const { geoLocations } = (await apolloClient.query({
      query: geoLocationsQuery,
      variables: { ipAddresses },
    })).data

    // Join the data about the location of the honey badger with the existing data about honey
    // badgers
    geoLocations.forEach(({ ipAddress, ...rest }) => {
      const honeyBadger = honeyBadgers.get(ipAddress)

      honeyBadgers.set(ipAddress, { ...honeyBadger, geoLocation: { ...rest } })
    })
  }

  /**
   * Creates a new stateful React component that is responsible for providing the "root" for the
   * visual and logical React components that make up the app.
   */
  constructor() {
    super()

    this.state = initialState
  }

  /**
   * React component lifecycle method that will be called after the <App> component has been mounted
   * into the DOM. After which we make calls to our GraphQL service/API endpoint to collect
   * information about honey badgers and update our state with it.
   */

  // It is okay to call 'setState()' in 'componentDidMount()' as discussed in the React docs
  /* eslint-disable react/no-did-mount-set-state */
  async componentDidMount() {
    try {
      // Query for the top honey badgers from our GraphQL service/API endpoint and enqueue a request
      // to set the apps state to the honey badgers
      const honeyBadgers = await App.collectTopHoneyBadgersInfo()

      this.setState({ honeyBadgers })

      // Now collect the following info about honey badgers from our GraphQL service/API endpoint
      // and enqueue a request to set the state of the app with the new info we collected:
      // * The autonomous systems that oversee the IP address of the honey badger
      // * The geospatial location of a honey badgers IP address
      await App.collectAutonomousSystemsInfo(honeyBadgers)
      this.setState({ honeyBadgers })

      await App.collectGeoLocationsInfo(honeyBadgers)
      this.setState({ honeyBadgers })
    } catch (error) {
      // It is okay to call 'setState()' in 'componentDidMount()' as discussed in the React docs
      // eslint-disable-next-line react/no-did-mount-set-state
      this.setState({ error })
    }
  }

  render() {
    const { honeyBadgers, error } = this.state

    if (error) {
      return `The following error occurred: ${error}`
    }

    return <HoneyBadgersTable honeyBadgers={[...honeyBadgers.values()]} />
  }
}

export default App
