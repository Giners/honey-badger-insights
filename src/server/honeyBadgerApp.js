import express from 'express'
import cors from 'cors'

import config from './config'
import apiService from './services/apiService'

/**
 * An Express application. Express is a minimalist web framework and is responsible for handling the
 * HTTP requests sent to a host server. It has already been configured to make use of 3rd-party
 * middleware to facilitate the processing of HTTP requests but also to make use of middleware to
 * expose the services of our app.
 *
 * A server can be started from this by invoking the 'listen()' method. It will start a UNIX socket
 * and listen for connections on a given path. This method is identical to Node's
 * 'http.Server.listen()' method.
 *
 * @example
 * const port = 3000
 *
 * honeyBadgerApp.listen(port, () => {
 *   console.log(`HoneyBadgerApp now listening on port ${port}`)
 * })
 */
const honeyBadgerApp = express()

// Use 'cors' middleware to enable all CORS requests
honeyBadgerApp.use(cors())

// Handle API requests to our app
honeyBadgerApp.use(config.serviceURIs.graphQL, apiService)

export default honeyBadgerApp
