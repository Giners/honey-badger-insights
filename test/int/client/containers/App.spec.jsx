/* eslint prefer-arrow-callback: 0, func-names: 0 */
/* eslint no-unused-expressions: 0, chai-friendly/no-unused-expressions: 2 */

// General 3rd-party supporting test libs
import { expect } from 'chai'
import { mount } from 'enzyme'
import waitUntil from 'async-wait-until'

// General 3rd-party supporting libs
import React from 'react'

// App React component under test
import App from './../../../../src/client/containers/App'

// Supporting app code
import config from './../../../../src/server/config'
import honeyBadgerApp from './../../../../src/server/honeyBadgerApp'
import HoneyBadgersTable from './../../../../src/client/components/HoneyBadgersTable'

describe('React component test: <App>', function() {
  // Reference to our server we use to test the integration of the <App> component with our
  // backend. Needs to be cleaned up when the tests end.
  let testServer = null

  before('Setup test server with GraphQL API service/endpoint', function(done) {
    testServer = honeyBadgerApp.listen(config.app.port, () => {
      expect(testServer).to.not.be.null

      done()
    })
  })

  after('Cleanup test server with GraphQL API service/endpoint', function(
    done,
  ) {
    testServer.close(() => {
      done()
    })
  })

  describe('Renders correctly for given application state:', function() {
    it('Honey badgers after querying the GraphQL API endpoint', async function() {
      // Increase the timeout in our integration test since we are ultimately making async calls to
      // HoneyDB
      this.timeout(15000)
      const appWrapper = mount(<App />)
      expect(appWrapper).to.not.be.null
      expect(appWrapper.exists()).to.be.true

      await waitUntil(
        () => {
          if (appWrapper.state('error') != null) {
            throw new Error(appWrapper.state('error'))
          }

          return appWrapper.state('honeyBadgers').size > 0
        },
        15000,
        500,
      )

      // Force a re-render since we are updating our state out-of-band in regards to this test (i.e.
      // <App> makes async calls to our backend to get the honey badgers)
      appWrapper.update()

      const totalHoneyBadgers = appWrapper.state('honeyBadgers').size

      // Check that we got some honey badgers and that we have rendered the appropriate amount
      expect(totalHoneyBadgers).to.be.above(0)
      expect(appWrapper.find(HoneyBadgersTable).exists()).to.be.true

      // If we passed the assertions above set the state to have one honey badger with the same IP
      // address before we snapshot our component. That is because the above assertions are against
      // dynamic data returned by our server which would cause future snapshot assertions to fail,
      // well because, they data is dynamic and is always changing.
      const honeyBadgers = new Map([
        ['1.1.1.1', { ipAddress: '1.1.1.1', count: 987 }],
      ])
      appWrapper.setState({ honeyBadgers })
    })
  })

  describe('collectTopHoneyBadgersInfo():', function() {
    it('Promise resolves to map with honey badgers info', async function() {
      const honeyBadgers = await App.collectTopHoneyBadgersInfo()

      // Check that we correctly populated our keys on our map... Something is going on that is
      // weird where I can't convert the array like object into an object to use 'forEach'. Thus I
      // just iterate over it using a good old for loop.
      const keys = [...honeyBadgers.keys()]
      for (let i = 0; i < keys.length; i += 1) {
        expect(keys[i]).to.be.a('string')
        expect(keys[i]).to.not.be.empty
      }

      // Check that we correctly populated our entries on our map...
      ;[...honeyBadgers.values()].forEach(value => {
        expect(value).to.be.an('object')
        expect(value.ipAddress).to.be.a('string')
        expect(value.ipAddress).to.not.be.empty
        expect(value.count).to.be.a('number')
        expect(value.count).to.be.above(0)
      })
    })
  })

  describe('collectAutonomousSystemsInfo():', function() {
    it('Promise resolves to map with autonomous systems info injected', async function() {
      // First query for honey badgers so we can pass them to the
      // 'App.collectAutnomousSystemsInfo()' method...
      let honeyBadgers = await App.collectTopHoneyBadgersInfo()

      // Now query for autonomous systems info and check that the map of honey badgers has been
      // updated inline with the invocation of the 'App.collectAutonomousSystemsInfo()' method...
      await App.collectAutonomousSystemsInfo(honeyBadgers)

      // Check that we correctly populated our entries on our map...
      honeyBadgers = [...honeyBadgers.values()]
      honeyBadgers.forEach(honeyBadger => {
        expect(honeyBadger).to.be.an('object')
        expect(honeyBadger.as).to.be.an('object')

        const { name, asn, countryCode } = honeyBadger.as

        expect(name).to.be.a('string')
        expect(name).to.not.be.empty
        expect(asn).to.be.a('number')
        expect(asn).to.be.above(0)
        expect(countryCode).to.be.a('string')
        expect(countryCode).to.not.be.empty
      })
    })
  })

  describe('collectGeoLocationsInfo():', function() {
    it('Promise resolves to map with geospatial locations info injected', async function() {
      // First query for honey badgers so we can pass them to the
      // 'App.collectGeoLocationsInfo()' method...
      let honeyBadgers = await App.collectTopHoneyBadgersInfo()

      // Now query for geospatial locations info and check that the map of honey badgers has been
      // updated inline with the invocation of the 'App.collectGeoLocationsInfo()' method...
      await App.collectGeoLocationsInfo(honeyBadgers)

      // Check that we correctly populated our entries on our map...
      honeyBadgers = [...honeyBadgers.values()]
      honeyBadgers.forEach(honeyBadger => {
        expect(honeyBadger).to.be.an('object')
        expect(honeyBadger.geoLocation).to.be.an('object')

        const {
          latitude,
          longitude,
          country,
          continent,
        } = honeyBadger.geoLocation

        expect(latitude).to.be.a('number')
        expect(longitude).to.be.a('number')
        expect(country).to.be.a('string')
        expect(country).to.not.be.empty
        expect(continent).to.be.a('string')
        expect(continent).to.not.be.empty
      })
    })
  })
})
