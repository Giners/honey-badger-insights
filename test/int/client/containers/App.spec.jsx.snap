// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React component test: <App> Renders correctly for given application state: Honey badgers after querying the GraphQL API endpoint 1`] = `
<App>
  <WithStyles(List)>
    <List
      classes={
        Object {
          "dense": "MuiList-dense-3",
          "padding": "MuiList-padding-2",
          "root": "MuiList-root-1",
          "subheader": "MuiList-subheader-4",
        }
      }
      component="ul"
      dense={false}
      disablePadding={false}
    >
      <ul
        className="MuiList-root-1 MuiList-padding-2"
      >
        <WithStyles(ListItem)
          key="1.1.1.1"
        >
          <ListItem
            ContainerComponent="li"
            button={false}
            classes={
              Object {
                "button": "MuiListItem-button-13",
                "container": "MuiListItem-container-6",
                "default": "MuiListItem-default-8",
                "dense": "MuiListItem-dense-9",
                "disabled": "MuiListItem-disabled-10",
                "divider": "MuiListItem-divider-11",
                "focusVisible": "MuiListItem-focusVisible-7",
                "gutters": "MuiListItem-gutters-12",
                "root": "MuiListItem-root-5",
                "secondaryAction": "MuiListItem-secondaryAction-14",
              }
            }
            dense={false}
            disableGutters={false}
            disabled={false}
            divider={false}
          >
            <li
              className="MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12"
              disabled={false}
            >
              <WithStyles(ListItemText)
                key=".0"
                primary="1.1.1.1"
              >
                <ListItemText
                  classes={
                    Object {
                      "dense": "MuiListItemText-dense-17",
                      "inset": "MuiListItemText-inset-16",
                      "primary": "MuiListItemText-primary-18",
                      "root": "MuiListItemText-root-15",
                      "secondary": "MuiListItemText-secondary-19",
                      "textDense": "MuiListItemText-textDense-20",
                    }
                  }
                  disableTypography={false}
                  inset={false}
                  primary="1.1.1.1"
                >
                  <div
                    className="MuiListItemText-root-15"
                  >
                    <WithStyles(Typography)
                      className="MuiListItemText-primary-18"
                      variant="subheading"
                    >
                      <Typography
                        align="inherit"
                        className="MuiListItemText-primary-18"
                        classes={
                          Object {
                            "alignCenter": "MuiTypography-alignCenter-34",
                            "alignJustify": "MuiTypography-alignJustify-36",
                            "alignLeft": "MuiTypography-alignLeft-33",
                            "alignRight": "MuiTypography-alignRight-35",
                            "body1": "MuiTypography-body1-30",
                            "body2": "MuiTypography-body2-29",
                            "button": "MuiTypography-button-32",
                            "caption": "MuiTypography-caption-31",
                            "colorError": "MuiTypography-colorError-44",
                            "colorInherit": "MuiTypography-colorInherit-40",
                            "colorPrimary": "MuiTypography-colorPrimary-41",
                            "colorSecondary": "MuiTypography-colorSecondary-42",
                            "colorTextSecondary": "MuiTypography-colorTextSecondary-43",
                            "display1": "MuiTypography-display1-25",
                            "display2": "MuiTypography-display2-24",
                            "display3": "MuiTypography-display3-23",
                            "display4": "MuiTypography-display4-22",
                            "gutterBottom": "MuiTypography-gutterBottom-38",
                            "headline": "MuiTypography-headline-26",
                            "noWrap": "MuiTypography-noWrap-37",
                            "paragraph": "MuiTypography-paragraph-39",
                            "root": "MuiTypography-root-21",
                            "subheading": "MuiTypography-subheading-28",
                            "title": "MuiTypography-title-27",
                          }
                        }
                        color="default"
                        gutterBottom={false}
                        headlineMapping={
                          Object {
                            "body1": "p",
                            "body2": "aside",
                            "display1": "h1",
                            "display2": "h1",
                            "display3": "h1",
                            "display4": "h1",
                            "headline": "h1",
                            "subheading": "h3",
                            "title": "h2",
                          }
                        }
                        noWrap={false}
                        paragraph={false}
                        variant="subheading"
                      >
                        <h3
                          className="MuiTypography-root-21 MuiTypography-subheading-28 MuiListItemText-primary-18"
                        >
                          1.1.1.1
                        </h3>
                      </Typography>
                    </WithStyles(Typography)>
                  </div>
                </ListItemText>
              </WithStyles(ListItemText)>
            </li>
          </ListItem>
        </WithStyles(ListItem)>
      </ul>
    </List>
  </WithStyles(List)>
</App>
`;
